File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(root==NULL) return root;
        if(root->val==p->val or root->val==q->val) return root;
        if(root->val>p->val && root->val<q->val || root->val<p->val && root->val>q->val) return root;
        if(root->val<p->val && root->val<q->val) return lowestCommonAncestor(root->right,p,q);
        else return lowestCommonAncestor(root->left,p,q);
    }
};

File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
var lowestCommonAncestor = function(node, p, q) {

        if(node.val===p.val || node.val===q.val){
            return node;
        }
        if(node.val<p.val && node.val>q.val || node.val<q.val && node.val>p.val){
            return node;
        }
        if(node.val<p.val && node.val<q.val){
            return lowestCommonAncestor(node.right,p,q);
        }
        else{
            return lowestCommonAncestor(node.left,p,q);
        }
};